nemo/collections/common/tokenizers/aggregate_tokenizer.py

```diff
@@ -0,0 +1,149 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+from nemo.utils import logging
+
+__all__ = ['AggregateTokenizer']
+
+
+class DummyTokenizer:
+    def __init__(self, vocab):
+        self.vocab = vocab
+        self.vocab_size = len(vocab)
+
+    # this one is for the crowd
+    def get_vocab(self):
+        return self.vocab
+
+
+class AggregateTokenizer(TokenizerSpec):
+    '''
+    AggregateTokenizer, allowing one to combine existing tokenizers into one.
+        Args:
+        tokenizers: dict of tokenizers, keys are lang ids
+    '''
+
+    def __init__(self, tokenizers: Dict):
+
+        self.tokenizers_dict = tokenizers
+        self.vocabulary = []
+
+        # the tokenizers should produce non-overlapping, ordered token ids
+        # keys are language ids
+        self.token_id_offset = {}
+
+        # keys are tokenizer numbers
+        self.token_id_offset_by_tokenizer_num = {}
+        offset = 0
+        i = 0
+        for lang, tokenizer in self.tokenizers_dict.items():
+            self.token_id_offset[lang] = offset
+            self.token_id_offset_by_tokenizer_num[i] = offset
+            offset += len(tokenizer.vocab)
+            i += 1
+
+        for tokenizer in self.tokenizers_dict.values():
+            self.vocabulary.extend(tokenizer.vocab)
+
+        self.vocab_size = len(self.vocabulary)
+        logging.info(f'Aggregate vocab size: {self.vocab_size}')
+
+        # for compatibility purposes only
+        self.tokenizer = DummyTokenizer(self.vocabulary)
+
+        # lookup tables to speed up token to text
+        self.offset_token_ids_by_token_id, self.tokenizers_by_token_id = self._calculate_offsets()
+
+    def _calculate_offsets(self):
+        offsets = {}
+        tokenizers = {}
+        cur_num = 0
+        tot = len(self.tokenizers_dict)
+        for id in range(len(self.vocabulary)):
+            off_id = id - list(self.token_id_offset.values())[cur_num]
+            if cur_num + 1 < tot:
+                if id >= list(self.token_id_offset.values())[cur_num + 1]:
+                    cur_num += 1
+                    off_id = id - list(self.token_id_offset.values())[cur_num]
+            offsets[id] = off_id
+            tokenizers[id] = list(self.tokenizers_dict.values())[cur_num]
+
+        return offsets, tokenizers
+
+    def text_to_tokens(self, text, lang_id):
```

Comment:
In my initial branch / implementation, there was a separate mix-in for aggregate tokenizers with this signature vs the original method, entailing the creation of a separate, multi-lang model classes using the mix-in.  We could go back to that. In this implementation, we pass *args instead of just text. Open to suggestions.

Reply:
resolving for now pending discussion